### PR TITLE
added command line parameter -pidfile; the pidfile is only generated …

### PR DESCRIPTION
**Pull request by ktsaou**
…if this parameter is passed; #151 #156 

@alonbl please check.

### CODE REVIEW

**Comment by ktsaou** [comment]
@alonbl I have chown'd the pid file to user netdata runs. I don't know if this ok or not.
Also, I have modified the openrc and systemd services to use the new command line parameter.

**Comment by alonbl** [comment]
> I have chown'd the pid file to user netdata runs. I don't know if this ok or not.
@ktsaou: you don't, the wrapper service will touch the file and chown before executing you.

File: .gitignore [hunk 1]
```diff
@@ -56,3 +56,4 @@ web/control.html
 web/datasource.css
 web/gadget.xml
 web/index_new.html
+system/netdata-openrc
```
Comment (alonbl) [inline .gitignore]: belongs to different patch, right?

File: netdata-installer.sh [hunk 2]
```diff
@@ -314,7 +314,7 @@ do
 
 	count=$((count + 1))
 
-	pid=$(cat /var/run/netdata/netdata.pid 2>/dev/null)
+	pid=$(cat /var/run/netdata/netdata.pid 2>/dev/null || cat /var/run/netdata.pid 2>/dev/null)
```
Comment (alonbl) [inline netdata-installer.sh]: why do you need both? what do you need it anyway? both openrc and systemd will not use it, so you cannot assume anything if missing.

File: netdata-installer.sh [hunk 3]
```diff
@@ -334,7 +334,7 @@ echo >&2
 # run netdata
 
 echo >&2 "Starting netdata..."
-run ${NETDATA_PREFIX}/usr/sbin/netdata "${@}"
+run ${NETDATA_PREFIX}/usr/sbin/netdata -pidfile /var/run/netdata.pid "${@}"
```
Comment (alonbl) [inline netdata-installer.sh]: I almost sure this script should be removed.

File: src/daemon.c [hunk 4]
```diff
@@ -96,9 +82,9 @@ int become_user(const char *username)
 		return -1;
 	}
 
-	if(chown(rundir, pw->pw_uid, pw->pw_gid) != 0) {
-		error("Cannot chown directory '%s' to user %s.", rundir, username);
-		return -1;
+	if(pidfile[0]) {
+		if(chown(pidfile, pw->pw_uid, pw->pw_gid) != 0)
```
Comment (alonbl) [inline src/daemon.c]: chown only if you drop privileges.

File: src/main.c [hunk 5]
```diff
@@ -44,7 +44,20 @@ void netdata_cleanup_and_exit(int ret)
 	netdata_exit = 1;
 	rrdset_save_all();
 	// kill_childs();
-	unlink(RUN_DIR "/netdata.pid");
+
+	if(pidfile[0]) {
+		if(unlink(pidfile) != 0)
+			error("Cannot unlink pidfile '%s'.", pidfile);
```
Comment (alonbl) [inline src/main.c]: this is expected if you drop privileges, should warn at worse case.

File: src/main.c [hunk 6]
```diff
@@ -44,7 +44,20 @@ void netdata_cleanup_and_exit(int ret)
...
+	}
+
+	if(pidfd != -1) {
```
Comment (alonbl) [inline src/main.c]: hmm... you should probably close it before trying to unlink.
anyway, why do you keep it opened?

File: system/netdata-openrc.in [hunk 7]
```diff
@@ -25,7 +25,7 @@ extra_started_commands="getconf"
 pidfile="/run/netdata/netdata.pid"
 command="${NETDATA_INSTALL_PATH}/usr/sbin/netdata"
 command_background="yes"
-command_args="${NETDATA_EXTRA_ARGS}"
+command_args="-pidfile ${pidfile} ${NETDATA_EXTRA_ARGS}"
```
Comment (alonbl) [inline system/netdata-openrc.in]: no... in openrc you do not pass this pidfile to application.

modify: pidfile="/var/run/netdata.pid"
remove: -pidfile ${pidfile}
add: start_stop_daemon_args="--user netdata:netdata"

add:
```
start_pre() {
	mkdir -p /var/cache/netdata
	chown netdata:netdata /var/cache/netdata
}
```

Comment (alonbl) [inline system/netdata-openrc.in]: this is .in file now, so instead of netdata and /var/cache use the @macros@

Comment (alonbl) [inline system/netdata-openrc.in]: oh... please also update command_args so that it won't fork.

File: system/netdata.service.in [hunk 8]
```diff
@@ -8,7 +8,7 @@ WorkingDirectory=/tmp
 User=root
 Group=root
 PIDFile=@localstatedir_POST@/run/netdata/netdata.pid
-ExecStart=@sbindir_POST@/netdata
+ExecStart=@sbindir_POST@/netdata -pidfile $PIDFile
```
Comment (alonbl) [inline system/netdata.service.in]: you do not need pidfile for systemd, just don't fork.

File: system/netdata.service.in [hunk 9]
```diff
@@ -8,7 +8,7 @@ WorkingDirectory=/tmp
 User=root
 Group=root
```
Comment (alonbl) [inline system/netdata.service.in]: inject netdata as user and group, you have @macro@ for this.

File: system/netdata.service.in [hunk 10]
```diff
@@ -8,7 +8,7 @@ WorkingDirectory=/tmp
...
 PIDFile=@localstatedir_POST@/run/netdata/netdata.pid
```
Comment (alonbl) [inline system/netdata.service.in]: remove ^

File: system/netdata.service.in (hunk 9 above)
Comment (alonbl) [inline system/netdata.service.in]: use type=simple and avoid forking

File: system/netdata.service.in [hunk 11]
```diff
@@ -8,7 +8,7 @@ WorkingDirectory=/tmp
...
-ExecStart=@sbindir_POST@/netdata
+ExecStart=@sbindir_POST@/netdata -pidfile $PIDFile
 ExecStop=/bin/kill -SIGTERM $MAINPID
```
Comment (alonbl) [inline system/netdata.service.in]: no need ^ once type is simple

File: .gitignore (hunk 1 above)
Comment (ktsaou) [inline .gitignore]: I have merged a PR from @jlec , who moved the files to `.in`

File: netdata-installer.sh (hunk 2 above)
Comment (ktsaou) [inline netdata-installer.sh]: transition for the ones having installed the old installer.

File: netdata-installer.sh (hunk 3 above)
Comment (ktsaou) [inline netdata-installer.sh]: sure, but we would have a thousand of issues without it.

File: src/main.c (hunk 6 above)
Comment (ktsaou) [inline src/main.c]: to truncate it. Otherwise, the truncation will fail too.

File: system/netdata-openrc.in (hunk 7 above)
Comment (ktsaou) [inline system/netdata-openrc.in]: I am lost... could you please do this?

**Comment by ktsaou** [comment]
@alonbl I am lost... Please make the `system/*.in` changes you suggest...

**Comment by ktsaou** [comment]
pidfile is now chown'd only when dropping privileges.
pidfile is now kept open only when dropping privileges and chown failed.

File: netdata-installer.sh (hunk 3 above)
Comment (alonbl) [inline netdata-installer.sh]: question: if one use systemd or openrc, should he use this script?

File: src/daemon.c [hunk 12]
```diff
@@ -293,15 +291,22 @@ int become_daemon(int dont_fork, int close_all_files, const char *user, const ch
 		close(dev_null);
 
 	// generate our pid file
-	{
+	if(pidfile[0]) {
 		unlink(pidfile);
```
Comment (alonbl) [inline src/daemon.c]: oh! just saw this... you must not unlink pidfile, as you do not have permission to create a new file.

File: src/main.c (hunk 6 above)
Comment (alonbl) [inline src/main.c]: first truncate, close then unlink?

but I still do not understand what is the value to keep the pidfile opened all the time?

File: netdata-installer.sh (hunk 3 above)
Comment (ktsaou) [inline netdata-installer.sh]: I am not sure I get the question.
The whole idea of this script is to help users install it quickly. If you look at the blogs, this little thing made a difference. When netdata will be in all distros, this script will be useless. So we need it just to let everyone install it in the mean time.

File: system/netdata-openrc.in (hunk 7 above)
Comment (alonbl) [inline system/netdata-openrc.in]: prefix should be:
```
extra_started_commands="getconf"
pidfile="/run/netdata.pid"
command="${NETDATA_INSTALL_PATH}/usr/sbin/netdata"
command_background="yes"
command_args="-nodaemon ${NETDATA_EXTRA_ARGS}"
start_stop_daemon_args="-u ${NETDATA_OWNER}"
```

then remove from depend:
```
checkpath -o ${NETDATA_OWNER} -d @localstatedir_POST@/cache/netdata /run/netdata
```

Add:
```
start_pre() {
    checkpath -o "${NETDATA_OWNER}" -d "@localstatedir_POST@/cache/netdata"
}
```

besides, I do not like the new script that downloads the configuration and waiting for termination, this is not normal behavior and should be removed in favor of a custom command, the getconf is a good example, the start_post and stop_post should be removed.

File: system/netdata.service.in (hunk 8 above)
Comment (alonbl) [inline system/netdata.service.in]: should be:
```
@sbindir_POST@/netdata -nodaemon
```

File: system/netdata.service.in (hunk 9 above)
Comment (alonbl) [inline system/netdata.service.in]: should be:
```
Type=simple
User=@NETDATA_USER@
```

and you should add to configure.ac subst of NETDATA_USER with ${with_user}, also fix the openrc script with this as default.

File: netdata-installer.sh (hunk 3 above)
Comment (alonbl) [inline netdata-installer.sh]: ok, so we agree that this script should not be used in any supported (packaged) system, so I just ignore it, it has too many issues anyway, such as not using the autoconf vars.

File: src/daemon.c (hunk 12 above)
Comment (ktsaou) [inline src/daemon.c]: I am confused again.

There are 2 cases regarding the pidfile:

1. Started are root
2. Started as normal user

Now, we have to decide regarding these:

i. are we going the create a new pidfile or we assume it is already created?
ii. are we going to chown, when we drop privileges?
iii. are we going to truncate it?
iv. are we going to unlink it?

So, can you give me a few guidelines?

PS: `unlink()` removes an entry (unlinks the file) from the directory tree. It has nothing to do with already open files. Open files can be operated as normal, although unlinked.

Comment (alonbl) [inline src/daemon.c]: 1. I know that unlink can be done on opened file, but this should be performed in unique cases.
2. if you run as netdata and pid is on /run then the packaging will touch /run/netdata.pid && chown netdata /run/netdata.pid then execute your program. this means that if you remove this file, you won't be able to create it as /run permissions do not allow that. this is why you should truncate before write, and fallback to truncate when cleanup.
3. end result: you have valid pid within file regardless if you started as root or netdata, and when stop, either have pid file removed or size 0.

Comment (ktsaou) [inline src/daemon.c]: ok

1. I removed the unlink() that was called at start
2. I also moved the unlink after the truncate

**Comment by ktsaou** [comment]
@alonbl are you fixing the system/ files or you expect me to do so?

**Comment by alonbl** [comment]
@ktsaou: you can squash all these commits... I guess you leave the openrc/systemd for later. I still do not understand the value of keeping pidfile open for entire run... but rest is ok... :)

**Comment by alonbl** [comment]
@ktsaou: regarding systemd/openrc, I cannot push into your branch (I think)... I can push ontop, so just leave these out we will have another commit later.

**Comment by ktsaou** [comment]
> I still do not understand the value of keeping pidfile open for entire run...

My bad. I believed `become_user()` is always called. I fixed it.

**Comment by ktsaou** [comment]
Well, I run `git merge --squash pidfile` and it merged it.

**Comment by JaviSabalete** [comment]
Today, I installed netdata on Debian Jessie 8.3 and I wrote a crappy script

[netdata.txt](https://github.com/firehol/netdata/files/207031/netdata.txt)

I'm not an expert, but works... enough for me :sweat_smile:
